README.md
# psychRstats

This is an R package based on Bard College's Statistics for Psychology (PSY 203, Fall 2020), taught by Assistant Professor of Psychology Justin Dainer-Best. It has tutorials and exercises to be used alongside an undergraduate course in statistics. (If you are familiar with statistical concepts and want to learn to perform such tests in R, it may be appropriate.)

Installing the package (see below) will download vignettes and exercises which build on one another to develop skills while also exploring basic statistical concepts like _t_-tests, ANOVA, and null-hypothesis statistical testing. (While there are many good arguments against NHST, I do not go into them in these tutorials.)

The primary website for that course was <https://faculty.bard.edu/~jdainerbest/psy-203/> with all labs being listed on [the lab website](https://faculty.bard.edu/~jdainerbest/psy-203/labslist.html); the labs as used in the course are available at <https://github.com/jdbest/psy-203/>. This repository takes the `{learnr}` tutorials which were the basis of those labs and deploys them as package vignettes, meaning that you can use *this page* and the included website (<https://jdbest.github.io/psychRstats/>) for all exercises. 

## Installation

For information on installing R and RStudio, visit the website.

In R, you can install this package directly from github:

```
# install devtools for the following function, if needed
install.packages("devtools")
devtools::install_github("jdbest/psychRstats")
library(psychRstats)
```

If you're prompted to update packages, you should respond as prompted -- usually with a number (e.g., typing a 1 and then hitting enter). Alternatively, leave the line blank for "don't update."

## Running tutorials

Once installed, tutorials can be used to learn basic concepts. The basic syntax is as follows:

```
psychRstats::lab("01-intro-to-r")
```

Once you've gotten through the first tutorial, you'll want to download the exercises as well. 

## Downloading exercises

Download all exercises and data in one go when you first get started. After running the above code and installing the `{psychRstats}` package, make sure you're in the right working directory.

(We discuss setting a working directory [in the first lab](https://jdbest.github.io/psychRstats/labs/01-intro-to-r/#set-a-working-directory), but there's also an entry about it [on the wiki](https://github.com/jdbest/psychRstats/wiki/setting-a-working-directory).)

```
psychRstats::download_exercises()
```

## Get started

I recommend following the [labs in order](https://jdbest.github.io/psychRstats/labs.html). Following along through the web may be helpful for most individuals. (For example, there is no tutorial in labs 2 or 5.)

NEWS.md
# psychRstats: Updates

## 0.1.0, 01-29-2021

* Technically finished---every file that should be part of this is now part of it
* All tutorials are included; all exercises are included

## 0.0.5, 01-29-2021

* Added labs 5-10

## 0.0.4, 01-29-2021

* Changing repo name to match package
* Updating a few labs
* Removing references to course assistant and Brightspace

## 0.0.3, 01-09-2021

* Add more functions to simplify
* Included two more labs and a zip version of exercises

## 0.0.2, 01-08-2021

* Got package running and amended README
* Included info on how to install and how to download exercises
* Updated website on github pages to have both exercises and answers
* Included first tutorial

## 0.0.1, 12-03-2020

* Created package and beginning basic functionality

R/psychRstats.R
#' psychRstats: College-Level Intro Stats Exercises in R
#'
#' Based on Bard College's Statistics for Psychology (PSY 203, Fall
#' 2020, taught by Assistant Professor of Psychology Justin 
#' Dainer-Best), these vignettes and exercises are intended to 
#' be used alongside an undergraduate course in statistics, or 
#' an understanding of statistics based on such a course. 
#' It includes a series of vignettes and exercises which build on 
#' one another to develop skills while also exploring basic 
#' statistical concepts like the t-test, ANOVA, or null-hypothesis 
#' statistical testing. 
#'
#' @section psychRstats functions:
#' There are two functions. The first, \code{\link{lab}},
#' will run a tutorial. It is just a mask of 
#' \code{learnr::run_tutorial()} with the package defaulting to 
#' this one.
#' 
#' The second function is \code{\link{run_exercise}},
#' which will open an R Markdown exercise from this package for 
#' editing, assuming you are in the current working directory.
#' 
#' For more information, look at the main github site,
#' \url{https://github.com/jdbest/psychRstats/}
#'
#' @docType package
#' @name psychRstats
NULL




# --------------------------------------------------------------
#' Load a tutorial from the psychRstats package using packages
#' \code{learnr} and, where appropriate, \code{gradethis}.
#' 
#' @param name The name of the tutorial, in quotation marks. 
#'   A full list of tutorials can be found by running 
#'   \code{\link{list_tutorials}}
#' 
#' @examples 
#' \dontrun{
#' lab("01-intro-to-r")
#' lab(1)
#' }
#' 
#' @export
lab <- function(name) {
  if(missing(name)) stop("You must choose a lab to run, e.g., lab(\"01-intro-to-r\"). \nRun the function list_tutorials() if you would like a list of all possible psychRstats tutorials.\n")
  if(! is.character(name) & ! is.numeric(name)) stop("The lab tutorial name must be a character/string vector, e.g., lab(\"01-intro-to-r\"), or a number, e.g., lab(1). \n")
  if(is.character(name)) name <- tolower(name)
  if(is.numeric(name)) name <- tutorials[tutorials$number == name,]$name
  if(is.na(name)) stop("There is no tutorial for the lab you have requested.")
  learnr::run_tutorial(name, package = "psychRstats")
}


#-------------------------------------------------------------- 
#' Open an exercise for editing from the psychRstats package
#' 
#' Using the utils function \code{\link[utils]{file.edit}}, 
#' will open the file for 
#' editing based on the current working directory. 
#' Please note that if the file does not already exist, it 
#' will not be created; an error will be returned. In this case, 
#' you should search your computer for the correct directory, 
#' and consider changing R's working directory.
#' 
#' @param exercise The name or number of the exercise, e.g., 
#'   \code{run_exercise(3)} or \code{run_exercise("03-exercise")}
#' @param dir The folder/directory to look for the exercise. Defaults
#'   to your current working directory.
#' 
#' @examples
#' \dontrun{
#' run_exercise("03-exercise")
#' run_exercise(3)
#' }
#' 
#' @export
run_exercise <- function(exercise, dir) {
  if(missing(dir)) path <- getwd() else path <- file.path(dir)
  if(missing(exercise)) stop("You must choose an exercise to open, e.g., run_exercise(\"02-exercise\"). \nRefer to the website if you have questions.\n")
  if(! is.character(exercise) & ! is.numeric(exercise)) stop("The lab exercise name must be a character/string vector, e.g., run_exercise(\"02-exercise\"), or a number, e.g., run_exercise(2). \n")
  if(is.character(exercise)) {
    exercise <- base::tolower(exercise)
    if(grepl(pattern = ".rmd", x = exercise)) { # remove rmd if kept
      exercise <- unlist(strsplit(exercise, split = "\\."))[1]
    }
  }
  if(is.numeric(exercise)) {
    exercise <- exercises[exercises$number == exercise,]$name
  }
  # add exercises folder if missing
  if(file.exists(file.path(path, "exercises"))) {
    path <- file.path(path, "exercises")
  }
  # add folder for current exercise
  path <- file.path(path, exercise)
  if(!file.exists(path)) {
    message <- paste0("The directory indicated, ",
                     path, "/ doesn't exist.\n")
    stop(message)
  }
  # add filename
  path <- file.path(path, paste0(exercise, ".Rmd"))
  # code amended from usethis::edit_file() to either open file in
  # R Studio or to use file.edit() to open it
  if (rstudioapi::isAvailable() && 
      rstudioapi::hasFun("navigateToFile")) {
    rstudioapi::navigateToFile(path)
  } else {
    utils::file.edit(path)
  }
  invisible(path)
}



#-----------------------------------------------------
#' Print all available tutorials in the psychRstats package.
#' 
#' Include no arguments to print them all. You may pass an argument
#' to the print function, e.g., \code{list_tutorials(n = 2)} to 
#' limit the number of responses.
#' 
#' @param n How many rows to print.
#' 
#' @examples 
#' list_tutorials()
#' 
#' @export
list_tutorials <- function(n = NULL) {
  if(!is.null(n)) n <- n*2 # because print max is based on number
  # of entries, not number of rows...
  print(tutorials, max = n)
}


#-----------------------------------------------------
#' Download exercises for psychRstats
#' 
#' Wraps around the \code{\link[usethis]{use_zip}} function
#' to download the exercises. First checks whether the directory 
#' already exists. If you'd like to overwrite an existing directory,
#' you must use \code{force = TRUE}.
#' 
#' @param dir The directory defaults to your current working 
#'   directory, but can be set otherwise. 
#' @param force Whether to overwrite other folders if they exist
#'   -- defaults to FALSE, but can overwrite by being set to TRUE.
#' 
#' @examples 
#' \dontrun{
#' download_exercises()
#' download_exercises("~/Downloads/")
#' download_exercises(dir = "~/Downloads/", force = TRUE)
#' }
#' @export
download_exercises <- function(dir, force = FALSE) {
  if(missing(dir)) path <- getwd() else path <- file.path(dir)
  if(!file.exists(path)) {
    message <- paste("The directory indicated,",
                     path, "doesn't exist.\n")
    stop(message)
  }
  if(file.exists(file.path(path, "exercises")) & ! force) {
    message <- paste("There is already an exercises folder in the",
                     "directory ", path, 
                     "-- are you sure you want to overwrite?",
                     "If so, re-run with force = TRUE.\n")
    stop(message)
  }
  usethis::use_zip("https://github.com/jdbest/psychRstats/raw/main/exercises.zip", cleanup = TRUE, destdir = path)
}